unittest: Only add main.c if it exists

yperess opened the pull request:
Downstream it's easier to write tests that use target_sources after
including the unittest package instead of specifying a list of sources
before. But if we do that, currently, main.c is added and the build
breaks because it doesn't exist.

Commit:
unittest: Only add main.c if it exists

Downstream it's easier to write tests that use target_sources after
including the unittest package instead of specifying a list of sources
before. But if we do that, currently, main.c is added and the build
breaks because it doesn't exist.

Signed-off-by: Yuval Peress <peress@google.com>

tejlmand commented on cmake/modules/unittest.cmake:
```diff
@@ -33,7 +33,7 @@ if((NOT DEFINED ZEPHYR_BASE) AND (DEFINED ENV_ZEPHYR_BASE))
 endif()
 
-if(NOT SOURCES)
+if(NOT SOURCES AND EXISTS main.c)
```
we should probably deprecate the old style and solely rely on the user to do an `target_sources()`.

This stems back from CMake <3.11, where sources were a required argument to `add_executable()`.

But such cleanup should not prevent this improvement.

yperess replied:
100% agree, I guess we could add a warning for using `SOURCES` saying that the dev should add it themselves and we'll stop doing that soon.

tejlmand reviewed (dismissed):
Thanks for improving this.

Commit:
unittest: warn users of deprecated SOURCES

If users set SOURCES before find_package for unit tests, we'll warn them
that this code path will no longer be supported and provide the correct
way of doing things moving forward.

Signed-off-by: Yuval Peress <peress@google.com>

tejlmand commented on cmake/modules/unittest.cmake:
```diff
@@ -33,6 +33,14 @@ if((NOT DEFINED ZEPHYR_BASE) AND (DEFINED ENV_ZEPHYR_BASE))
 endif()
 
+if(SOURCES)
+  message(WARNING
+      "Unit testing will soon stop automatically adding SOURCES to testbinary. If not already done,"
+      " please use:\n"
+      "    target_sources(testbinary PRIVATE \${SOURCES})"
+  )
```
please move this to https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/FindDeprecated.cmake
along with this code:
https://github.com/zephyrproject-rtos/zephyr/blob/7ef05751a3f34030eb06dace23e357d10b33f460/cmake/modules/unittest.cmake#L36_L38
Plus a new `target_sources(testbinary PRIVATE ${SOURCES})` line, and then update code to use:

```
add_executable(testbinary)
find_package(Deprecated COMPONENTS SOURCES)
```

that allows us to have deprecated code located together in a single file and helps us to keep track and remove deprecated code after two release.

Plus, it has the benefit of already being able to clean up existing code :)

Note: align the warning message to similar messages in the deprecation module.

tejlmand reviewed (changes requested):
Thanks for the improvement, but before marking this deprecated and introducing the warning then we must minimum update all files inside the Zephyr repo to use the new pattern.

That would be those files:
```
tests/unit/cbprintf/CMakeLists.txt
tests/unit/util/CMakeLists.txt
tests/unit/crc/CMakeLists.txt
tests/unit/net_timeout/CMakeLists.txt
tests/unit/base64/CMakeLists.txt
tests/unit/list/CMakeLists.txt
tests/unit/winstream/CMakeLists.txt
tests/unit/math_extras/CMakeLists.txt
tests/unit/timeutil/CMakeLists.txt
tests/unit/intmath/CMakeLists.txt
tests/unit/rbtree/CMakeLists.txt
tests/ztest/base/CMakeLists.txt
tests/bluetooth/controller/ctrl_encrypt/CMakeLists.txt
tests/bluetooth/controller/ctrl_min_used_chans/CMakeLists.txt
tests/bluetooth/controller/ctrl_tx_buffer_alloc/CMakeLists.txt
tests/bluetooth/controller/ctrl_tx_queue/CMakeLists.txt
tests/bluetooth/controller/ctrl_invalid/CMakeLists.txt
tests/bluetooth/controller/ctrl_terminate/CMakeLists.txt
tests/bluetooth/controller/ctrl_data_length_update/CMakeLists.txt
tests/bluetooth/controller/ctrl_conn_update/CMakeLists.txt
tests/bluetooth/controller/ctrl_api/CMakeLists.txt
tests/bluetooth/controller/ctrl_unsupported/CMakeLists.txt
tests/bluetooth/controller/ctrl_chmu/CMakeLists.txt
tests/bluetooth/controller/ctrl_le_ping/CMakeLists.txt
tests/bluetooth/controller/ctrl_cte_req/CMakeLists.txt
tests/bluetooth/controller/ctrl_phy_update/CMakeLists.txt
tests/bluetooth/controller/ctrl_version/CMakeLists.txt
tests/bluetooth/controller/ctrl_feature_exchange/CMakeLists.txt
tests/bluetooth/controller/ctrl_hci/CMakeLists.txt
tests/bluetooth/controller/ctrl_collision/CMakeLists.txt
```

```

yperess commented:
> Thanks for the improvement, but before marking this deprecated and introducing the warning then we must minimum update all files inside the Zephyr repo to use the new pattern.
> 
> That would be those files:
> ```
> tests/unit/cbprintf/CMakeLists.txt
> tests/unit/util/CMakeLists.txt
> tests/unit/crc/CMakeLists.txt
> tests/unit/net_timeout/CMakeLists.txt
> tests/unit/base64/CMakeLists.txt
> tests/unit/list/CMakeLists.txt
> tests/unit/winstream/CMakeLists.txt
> tests/unit/math_extras/CMakeLists.txt
> tests/unit/timeutil/CMakeLists.txt
> tests/unit/intmath/CMakeLists.txt
> tests/unit/rbtree/CMakeLists.txt
> tests/ztest/base/CMakeLists.txt
> tests/bluetooth/controller/ctrl_encrypt/CMakeLists.txt
> tests/bluetooth/controller/ctrl_min_used_chans/CMakeLists.txt
> tests/bluetooth/controller/ctrl_tx_buffer_alloc/CMakeLists.txt
> tests/bluetooth/controller/ctrl_tx_queue/CMakeLists.txt
> tests/bluetooth/controller/ctrl_invalid/CMakeLists.txt
> tests/bluetooth/controller/ctrl_terminate/CMakeLists.txt
> tests/bluetooth/controller/ctrl_data_length_update/CMakeLists.txt
> tests/bluetooth/controller/ctrl_conn_update/CMakeLists.txt
> tests/bluetooth/controller/ctrl_api/CMakeLists.txt
> tests/bluetooth/controller/ctrl_unsupported/CMakeLists.txt
> tests/bluetooth/controller/ctrl_chmu/CMakeLists.txt
> tests/bluetooth/controller/ctrl_le_ping/CMakeLists.txt
> tests/bluetooth/controller/ctrl_cte_req/CMakeLists.txt
> tests/bluetooth/controller/ctrl_phy_update/CMakeLists.txt
> tests/bluetooth/controller/ctrl_version/CMakeLists.txt
> tests/bluetooth/controller/ctrl_feature_exchange/CMakeLists.txt
> tests/bluetooth/controller/ctrl_hci/CMakeLists.txt
> tests/bluetooth/controller/ctrl_collision/CMakeLists.txt
> ```
> 
> ```

I'll do that this weekend

Commit:
test: update how unit tests set sources

Replaces cases of setting SOURCES before calling find_package with
proper target_sources.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
bluetooth: fix host tests

Host tests were failing for -p unit_testing since FFF globals were
declared in multiple .c files. This caused a linker error for
multiple definitions.

Signed-off-by: Yuval Peress <peress@google.com>

yperess commented:
@aaronemassey @tejlmand any chance I can get a quick review on this?

alwa-nordic commented:
Rebase to get https://github.com/zephyrproject-rtos/zephyr/pull/51633. That replaces your last commit.

theob-pro commented:
https://github.com/zephyrproject-rtos/zephyr/blob/e9bf334d99c1f6676e7c5cd1bd90ef7b5a0b0e20/tests/bluetooth/host/keys/bt_keys_foreach_bond/CMakeLists.txt#L5

This should be changed as well as the other files in the `keys` folder.

tejlmand commented on tests/unit/util/CMakeLists.txt:
```diff
@@ -3,5 +3,5 @@
 project(util)
-set(SOURCES main.c maincxx.cxx ../../../lib/os/dec.c)
 find_package(Zephyr COMPONENTS unittest REQUIRED HINTS $ENV{ZEPHYR_BASE})
+target_sources(testbinary PRIVATE main.c maincxx.cxx ../../../lib/os/dec.c)
```
the rework in this PR allows you to use `ZEPHYR_BASE` instead of `../../../`

```suggestion
target_sources(testbinary PRIVATE main.c maincxx.cxx ${ZEPHYR_BASE}/lib/os/dec.c)
```

yperess replied:
Done

tejlmand commented on cmake/modules/FindDeprecated.cmake:
```diff
@@ -73,5 +73,12 @@ if(NOT "${Deprecated_FIND_COMPONENTS}" STREQUAL "")
 
+if("SOURCES" IN_LIST Deprecated_FIND_COMPONENTS)
+  message(DEPRECATION
+      "Setting SOURCES prior to calling find_package() for unit tests is deprecated."
+      "Please set SOURCES after find_package and use:\n"
+      "    target_sources(testbinary PRIVATE \${SOURCES})")
```
We should only ask users to call `target_sources()`, whether or not they decide to create a `SOURCES` variable for this purpose is up to them.

How about
```suggestion
      "Setting SOURCES prior to calling find_package() for unit tests is deprecated."
      "Please add sources after find_package and use:\n"
      "    target_sources(testbinary PRIVATE <source-file.c>)")
```

tejlmand reviewed (changes requested):
Those files must be update as well regarding `SOURCES`
```
./bluetooth/host/keys/bt_keys_foreach_type/CMakeLists.txt
./bluetooth/host/buf/bt_buf_get_rx/CMakeLists.txt
./bluetooth/host/buf/bt_buf_get_cmd_complete/CMakeLists.txt
./bluetooth/host/buf/bt_buf_get_evt/CMakeLists.txt
./bluetooth/host/buf/bt_buf_get_type/CMakeLists.txt
./bluetooth/controller/ctrl_cte_req/CMakeLists.txt
```

tejlmand reviewed (approved):
lgtm, thanks a lot for this cleanup
